Sources/NIOCore/AsyncSequences/NIOBackPressuredAsyncSequence.swift

```diff
@@ -0,0 +1,1003 @@
+                    self.delegate?.didTerminate()
+                    self.delegate = nil
+
+                    return
+
+                case .none:
+                    return
+                }
+            }
+        }
+
+        @inlinable
+        /* fileprivate */ internal func yield<S: Sequence>(_ sequence: S) -> Source.YieldResult where S.Element == Element {
+            self.lock.withLock {
+                let action = self.stateMachine.yield(sequence)
+
+                switch action {
+                case .returnDemandMore:
+                    return .demandMore
+
+                case .returnStopDemanding:
+                    return .stopDemanding
+
+                case .returnDropped:
+                    return .dropped
+
+                case .resumeContinuationAndReturnDemandMore(let continuation, let element):
+                    continuation.resume(returning: element)
+
+                    return .demandMore
+
+                case .resumeContinuationAndReturnStopDemanding(let continuation, let element):
+                    continuation.resume(returning: element)
+
+                    return .stopDemanding
+                }
+            }
+        }
+
+        @inlinable
+        /* fileprivate */ internal func finish() {
+            self.lock.withLock {
+                let action = self.stateMachine.finish()
+
+                switch action {
+                case .callDidTerminate:
+                    self.delegate?.didTerminate()
+                    self.delegate = nil
+
+                    return
+
+                case .resumeContinuationWithNilAndCallDidTerminate(let continuation):
+                    self.delegate?.didTerminate()
+                    self.delegate = nil
+                    continuation.resume(returning: nil)
+
+                    return
+
+                case .none:
+                    return
+                }
+            }
+        }
+
+        @inlinable
+        /* fileprivate */ internal func next() async -> Element? {
+            // We are locking manually here since we want to hold the lock
+            // across the suspension point.
+            self.lock.lock()
+
+            let action = self.stateMachine.next()
+
+            switch action {
+            case .returnElement(let element):
+                self.lock.unlock()
+                return element
+
+            case .returnElementAndCallDemand(let element):
+                self.delegate?.demand()
+                self.lock.unlock()
+
+                return element
+
+            case .returnElementAndCallDidTerminate(let element):
+                self.delegate?.didTerminate()
+                self.delegate = nil
+                self.lock.unlock()
+
+                return element
+
+            case .returnNil:
+                self.lock.unlock()
+                return nil
+
+            case .suspendTask:
+                // It is safe to hold the lock across this method
```

Comment:
that isn't true here; the cancellation case will happen in two ways; either a) it will cancel asynchronously from another task (which is safe) or b) it will cancel immediately because the task is already cancelled at the point of entry (which is unsafe because you will become re-rentrant upon the lock)

Reply:
I was super unsure about this since the documentation is a bit bogus about it. It states both

> This differs from the operation cooperatively checking for cancellation and reacting to it in that the cancellation handler is always and immediately invoked when the task is canceled

and
> Doesn’t check for cancellation, and always executes the passed operation.

From what I understood there is that the operation is at least always executed. What you are saying is that if the `Task` is cancelled then `operation` is never called and this will result in a deadlock here, right?

Reply:
Okay I just tested the following

```swift
func main() async {
    do {
        try await  Task.sleep(nanoseconds: 2000000000)
    } catch {
        print("got cancelled")
        await withTaskCancellationHandler {
            print("operation")
        } onCancel: {
            print("cancelled")
        }

    }
}

let task = Task {
    await main()
}

Task {
    try! await Task.sleep(nanoseconds: 1000000000)
    print("cancelling")
    task.cancel()
}
```

and this prints

```
cancelling
got cancelled
cancelled
operation
```

Which is what you probably meant here. Thanks for spotting this! Now the question is how this should be done instead :D

Reply:
Hm, thinking out loud about this a bit. Is checking for `Task.isCancelled` before enough or does that introduce a race condition.

Reply:
Thanks for catching this @phausler, I moved the `withTaskCancellationHandler` around the whole method body now. This should make sure that only either of the two gets to run without lock re-entrancy.